src/rule.rs

```diff
@@ -133,12 +153,44 @@ pub struct RuleSubvolume {
 }
 
 impl RuleSubvolume {
-    fn snapshot(&self) -> anyhow::Result<()> {
+    fn snapshot(&self, rule: &Rule) -> anyhow::Result<()> {
         let mut name = RandomName::new();
         for _ in 0..16 {
             let target_path = self.target_dir.join(name.as_str());
+            if !rule.pre_snapshot.is_empty() {
+                let mut cmd = std::process::Command::new("sh");
+                cmd.arg("-c").arg(format!("{} {} {}", rule.pre_snapshot, self.path.display(), target_path.display()));
+                let res = cmd.output();
+                match res {
+                    Ok(output) => {
+                        if !output.status.success() {
+                            let error_code = output.status.code();
+                            if let Some(code) = error_code {
+                                eprintln!("Pre Snapshot script failed, aborting snapshot, exit code: {}, error: {}", code, String::from_utf8_lossy(&output.stderr));
+                            } else {
+                                eprintln!("Pre Snapshot script failed, aborting snapshot, error: {}", String::from_utf8_lossy(&output.stderr));
+                            }
+                            return Err(anyhow::anyhow!("Pre snapshot script failed"));
+                        }
+                    }
+                    Err(e) => {
+                        eprintln!("failed to run Pre Snapshot script: {}", e);
+                        return Err(e.into());
+                    }
+                }
+            }
             match create_butter_snapshot(&self.path, &target_path, true) {
-                Ok(_) => return Ok(()),
+                Ok(_) => {
+                    if !rule.post_snapshot.is_empty() {
+                        let mut cmd = std::process::Command::new("sh");
+                        cmd.arg("-c").arg(format!("{} {} {}", rule.post_snapshot, self.path.display(), target_path.display()));
```

Comment:
Post snapshot hook here should be getting the path to the snapshot.
Taking custom command with support for placeholders for the variables we insert could be useful for quick things but I don't think it would make much difference right now.